graspy/embed/casc.py

```diff
@@ -0,0 +1,481 @@
+                    return res
+                
+
+            else:
+                def matmult(x, y,t=False):
+                    if t:
+                        res=np.dot(graphMat,np.dot(graphMat.transpose(),y))+\
+                        hTuningParam*np.dot( covariates, np.dot(covariates.transpose(), y))
+                    else:
+                        res=np.dot(graphMat,np.dot(graphMat,y)) + hTuningParam*np.dot(
+                                covariates, np.dot(covariates.transpose(), y))
+                    return res
+            
+                
+            
+            u2,d2,v2,it,pr = self.irlb(graphMat,nBlocks+1,matmult)	
+        
+        eVec = u2[:, 0:nBlocks]
+        eVal = d2[0:(nBlocks+1)]
+    
+        return {'eVec':eVec,'eVal':eVal}
+
+    def getCascResults(self,graphMat, covariates, hTuningParam,nBlocks, rowNorm,assortative,CCA):
+        cascSvd = self.getCascSvd(graphMat, covariates, hTuningParam, nBlocks, assortative,CCA)
+        
+        if rowNorm:
+            cascSvd_tmp = cascSvd['eVec']/np.sqrt(sum(cascSvd['eVec']**2))
+        else:
+            cascSvd_tmp = cascSvd['eVec']
+        kmeansResults=KMeans(n_clusters=nBlocks).fit(cascSvd_tmp)
+        
+        if not CCA:
+            ortho = self.getOrtho(graphMat, covariates,
+                                  cascSvd['eVec'], cascSvd['eVal'],hTuningParam, nBlocks)
+
+            return {'orthoL':ortho[0],'orthoX':ortho[1],
+                    'wcss':kmeansResults.inertia_,'cluster':kmeansResults.labels_,'cascSvd':cascSvd_tmp}
+    
+        else:
+            return {'orthoL':None,'orthoX':None,
+                    'wcss':kmeansResults.inertia_,'cluster':kmeansResults.labels_,'cascSvd':cascSvd_tmp}
+    
+    
+    
+    def mult(self,A, x, t=False):
+        assert x.ndim == 1
+        if(sp.issparse(A)):
+            if(t):
+                return(sp.csr_matrix(x).dot(A).transpose().todense().A[:, 0])
+            return(A.dot(sp.csr_matrix(x).transpose()).todense().A[:, 0])
+        if(t):
+            return np.asarray(A.transpose().dot(x)).ravel()
+        return np.asarray(A.dot(x)).ravel()
+    
+    def orthog(self,Y, X):
+        """Orthogonalize a vector or matrix Y against the columns of the matrix X.
+        This function requires that the column dimension of Y is less than X and
+        that Y and X have the same number of rows.
+        """
+        dotY = Y.dot(X)
+        return (Y - self.mult(X, dotY))
+    
+    # Simple utility function used to check linear dependencies during computation:
+    def invcheck(self,x):
+        eps2 = 2 * np.finfo(np.float).eps
+        if(x > eps2):
+            return 1.0 / x
+        else:
+            warnings.warn(
+                "Ill-conditioning encountered, result accuracy may be poor")
+            return 0.0
+    
+    def irlb(self,A, n,matmult,tol=0.0001, maxit=50, center=None, scale=None, random_state=0):

```

Comment:
cant just copy and paste this in, should not go in this file. Their code may work but it needs a lot of changes to style and the docstring in order to match the formatting for graspy. this should be in the graspy svd file

Reply:
I understand there are formatting issues here, will fix them .
But I don't think we should put this into SVD because it's not a generalized SVD method. It has the specific optimization for CASC within it.
If you still insist it should be seperate file in SVD, will do that